src/components/WalletSelector/walletUtils.ts

```diff
@@ -0,0 +1,205 @@
+import { JsonRpcSigner, Web3Provider } from '@ethersproject/providers';
+import GNOSIS_SAFE_CONTRACT_ABI from 'abis/gnosis-safe-contract.json';
+import { keccak256 } from '@ethersproject/keccak256';
+import { toUtf8Bytes } from '@ethersproject/strings';
+import { CONTRACT_ACCOUNT, EXTERNALLY_OWNED_ACCOUNT } from 'types/Wallet';
+import { AbstractConnector } from '@web3-react/abstract-connector';
+import { WalletConnectConnector } from '@web3-react/walletconnect-connector';
+import walletlinkSigner from './walletlinkSigner';
+import { Contract } from '@ethersproject/contracts';
+import { WalletLinkConnector } from '@web3-react/walletlink-connector';
+import { Web3Error } from 'types/Error';
+
+// The hard coded value that Gnosis Safe contract's isValidSignature method returns if the message was signed
+// https://github.com/gnosis/safe-contracts/blob/dec13f7cdab62056984343c4edfe40df5a1954dc6/contracts/handler/CompatibilityFallbackHandler.sol#L19
+const GNOSIS_VALID_SIGNATURE_MAGIC_VALUE = '0x1626ba7e';
+const GNOSIS_SAFE_SIGN_MESSAGE_EVENT_NAME = 'SignMsg';
+
+type EthereumAccountType = typeof EXTERNALLY_OWNED_ACCOUNT | typeof CONTRACT_ACCOUNT;
+type Signature = string;
+// The contract accounts that Gallery supports (Gnosis, Argent, etc). ID is used by the backend to know which contract to call
+type ContractAccount = { name: string; id: number };
+
+// Externally Owned Accounts will always have Wallet Type ID = 0 because the backend will handle all EOAs the same way.
+export const DEFAULT_WALLET_TYPE_ID = 0;
+export const GNOSIS_SAFE_WALLET_TYPE_ID = 1;
+// List of contract account wallets that Gallery supports. Used to detect if the user is trying to log in with a contract account
+const SUPPORTED_CONTRACT_ACCOUNTS: [ContractAccount] = [
+  { name: 'Gnosis Safe', id: GNOSIS_SAFE_WALLET_TYPE_ID },
+];
+
+/**
+ * Checks what type of Ethereum account the wallet is and signs the message accordingly
+ */
+export async function signMessage(
+  address: string,
+  nonce: string,
+  connector: AbstractConnector,
+  signer: JsonRpcSigner,
+  library?: Web3Provider
+): Promise<Signature> {
+  const accountType = getEthereumAccountType(connector);
+
+  if (accountType === CONTRACT_ACCOUNT) {
+    // if theres a nonce in local storage, prompt to try again
```

Comment:
is this a TODO, or is that what's actually happening here?

Reply:
good call, this is a todo for optimization